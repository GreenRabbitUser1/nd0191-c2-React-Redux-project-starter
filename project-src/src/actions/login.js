import * as API from '../../_DATA';

export const LOGIN_REQUEST = 'LOGIN_REQUEST';
export const LOGIN_FAILURE = 'LOGIN_FAILURE';
export const LOGIN_ERROR = 'LOGIN_ERROR';
export const LOGIN_SUCCESS = 'LOGIN_SUCCESS';
export const SESSION_RESTORE = 'SESSION_RESTORE';
export const ATTEMPT_RESTORE = 'ATTEMPT_RESTORE';

export function setUser(user){
    return {
        type: LOGIN_REQUEST,
        user
    };
}

export function attemptLogin({username, password}){
    return async (dispatch) => {
        dispatch({
            type: LOGIN_REQUEST
        });
        try {

            const {user, getUserError} = await getUser(username, password);
            if (getUserError && getUserError !== null){
                dispatch(getUserError);
                return false;
            }

            //  SUCCESS! Username and password matches, user has been authenticated
            let sessionToken = await fetch('/api/auth/generate-jwt', {
                body: JSON.stringify({username, password}),
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json'
                }
            });
            if (sessionToken.ok){
                sessionToken = await sessionToken.json();
                console.log('sessionToken:', sessionToken);
                localStorage.setItem('udacity-final-project-token', sessionToken);
                const lst = localStorage.getItem('udacity-final-project-token');
                console.info(lst);
                // console.info('lst: ', lst);
            }
            dispatch({
                type: LOGIN_SUCCESS,
                user
            });
            return true;

        } catch(error) {
            dispatch({
                type: LOGIN_FAILURE,
                error: error.message
            })
            return false;

        }
       
    }
}

async function getUser(username, password){
    let user;
    let error;
    const response = await API._getUsers();
    if (!response || response === null){
        error = {
            type: LOGIN_FAILURE,
            error: 'Did not find any users'
        };
    }
    try {
        user = response[username];
    } catch(e){}
    if (!user || user === null){
        error = {
            type: LOGIN_FAILURE,
            error: 'Did not find user matching this username'
        };
    }
    else {
        //  Check if the password matches
        if (user.password !== password){
            error = {
                type: LOGIN_FAILURE,
                error: 'Password did not match'
            };
        }
    }
    return {
        user,
        getUserError: error
    };
}

export function refreshUser(uId){
    return async (dispatch) => {
        console.log('Refreshing User ...');
        const response = await API._getUsers();
        let user;
        if (response && response !== null){
            user = response[uId];
            dispatch({
                type: LOGIN_SUCCESS,
                user
            });
            console.log('Refreshed.');
            return true;
        }
    }
}

export function attemptSessionRestore(){

    return async (dispatch) => {
        //  Check if the user has a logged in cookie
        let validSession = false;
        const sessionToken = localStorage.getItem('udacity-final-project-token');
        if (sessionToken && sessionToken !== null){
            console.log('There is a valid session token', sessionToken);
            let decodedToken = await fetch('/api/auth/decode-jwt', {
                body: JSON.stringify({
                    token: sessionToken
                }),
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json'
                }
            });
            if (decodedToken.ok){
                decodedToken = await decodedToken.json();
                //  With the decodedToken the following are now available: username, password, iat, exp
                //  Get the user, set the state in the store, redirect to / if on /login
                console.log('decodedToken', decodedToken);
                const {user, getUserError} = await getUser(decodedToken.username, decodedToken.password);
                if (getUserError && getUserError !== null){
                    dispatch(getUserError);
                    return validSession;
                }
                dispatch({
                    type: LOGIN_SUCCESS,
                    user
                });
                dispatch({
                    type: ATTEMPT_RESTORE
                });
                validSession = true;
            }
        }

        return validSession;

        // dispatch({
        //     type: SESSION_RESTORE
        // })
    }

}