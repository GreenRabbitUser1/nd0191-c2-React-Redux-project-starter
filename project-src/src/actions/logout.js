import * as API from '../../_DATA';

export const LOGOUT_USER = 'LOGOUT_USER';

export function attemptLogout(){
    return async (dispatch) => {
        localStorage.removeItem('udacity-final-project-token');
        console.log('Logged out');
        dispatch({
            type: LOGOUT_USER
        });
    }
}
