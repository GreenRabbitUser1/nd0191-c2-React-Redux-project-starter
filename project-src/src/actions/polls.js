export const ADD_POLL = 'ADD_POLL';
export const GET_POLLS = 'GET_POLLS';

import * as API from '../../_DATA.js';

export function addPoll(){
    return async (dispatch) => {
        //  No need to create route on server side since database/apis are in _DATA.js
        //  Make 
    }
}

export function getPolls(){
    return async (dispatch) => {
        const polls = await API._getQuestions();
        console.log('polls: ', polls);
        dispatch({
            type: GET_POLLS,
            polls
        });
    }
}