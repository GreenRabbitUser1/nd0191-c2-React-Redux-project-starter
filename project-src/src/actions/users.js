export const ADD_USER = 'ADD_USER';
export const GET_USERS = 'GET_USERS';
export const SAVE_USER = 'SAVE_USER';

import * as API from '../../_DATA.js';

export function getUsers(){
    return async (dispatch) => {
        const users = await API._getUsers();
        console.log('users: ', users);
        dispatch({
            type: GET_USERS,
            users
        });
    }
}

export function saveUser(_user, _new_data){
    return async (dispatch) => {
        const user = await API._saveUser(_user, _new_data)
        dispatch({
            type: SAVE_USER,
            user
        });
    }
}