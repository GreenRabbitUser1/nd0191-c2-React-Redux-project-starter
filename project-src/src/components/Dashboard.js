import { useEffect } from "react";
import { connect } from "react-redux";
import PollWidget from "./PollWidget";
import { IoMdEye } from "react-icons/io";
import { FaChevronDown } from "react-icons/fa";
import { useState } from "react";

const Dashboard = (props) => {

    const {polls, dispatch, user} = props;

    const [showViewOptions, setShowViewOptions] = useState(false);

    const [showUnansweredPolls, setShowUnansweredPolls] = useState(true);
    const [showAnsweredPolls, setShowAnsweredPolls] = useState(true);

    useEffect(() => {
        console.log('polls changed!', polls);
    }, [polls])

    const toggleShowViewOptions = () => {
        setShowViewOptions(!showViewOptions);
    }

    const toggleShowUnansweredPolls = () => {
        setShowUnansweredPolls(!showUnansweredPolls)
    }

    const toggleShowAnsweredPolls = () => {
        setShowAnsweredPolls(!showAnsweredPolls)
    }

    return (
        <>
            <div className="view-wrap">
                <div className="view-filter-top" onClick={toggleShowViewOptions}>
                    <IoMdEye />
                    <span>Filter View</span>
                    <FaChevronDown  className={`
                        view-filter-icon 
                        ${(showViewOptions
                            ? ' view-filter-open'
                            : ''
                        )}
                    `} />
                </div>
                {
                    showViewOptions &&
                    <div className="view-options-wrap">
                        <div className="view-options">
                            <div className="view-option">
                                <input type="checkbox" checked={showUnansweredPolls} onChange={toggleShowUnansweredPolls} />
                                <label>Unanswered Polls</label>
                            </div>
                            <div className="view-option">
                                <input type="checkbox" checked={showAnsweredPolls} onChange={toggleShowAnsweredPolls} />
                                <label>Answered Polls</label>
                            </div>
                        </div>
                    </div>
                }
            </div>
            <div className="polls">
                {
                    polls && showUnansweredPolls && 
                    <div className="poll-category">
                        <div className="poll-category-name">
                            Unanswered Polls
                        </div>
                        <div className="poll-wrap">
                            {
                                // Show the unanswered polls
                                
                                Object.keys(polls).map((p) => {
                                    if (!user.answers[p]){
                                        return (
                                            <PollWidget poll={polls[p]} key={p} />
                                        ) 
                                    }
                                })
                            }
                        </div>
                    </div>
                }
                {
                    polls && showAnsweredPolls && 
                    <div className="poll-category">
                        <div className="poll-category-name">
                            Answered Polls
                        </div>
                        <div className="poll-wrap">
                            {
                                // Show the answered polls
                                polls &&
                                Object.keys(polls).map((p) => {
                                    if (user.answers[p]){
                                        return (
                                            <PollWidget poll={polls[p]} key={p} />
                                        ) 
                                    }
                                })
                            }
                        </div>
                    </div>
                }
            </div>
        </>
    );
};

const mapStateToProps = (state) => ({
    polls: state.polls,
    user: state.user
});

export default connect(mapStateToProps)(Dashboard);