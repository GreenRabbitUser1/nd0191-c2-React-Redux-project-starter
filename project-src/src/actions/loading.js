import { SET_LOADING_START, SET_LOADING_END } from "../reducers/loading";

export function setLoadingStart(){
    return async (dispatch) => {
        console.log('Toggle Loading - Start');
        dispatch({
            type: SET_LOADING_START
        });
    }
}

export function setLoadingEnd(){
    return async (dispatch) => {
        console.log('Toggle Loading - End');
        dispatch({
            type: SET_LOADING_END
        });
    }
}