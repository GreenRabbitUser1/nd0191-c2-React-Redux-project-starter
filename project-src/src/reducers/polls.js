import { ADD_POLL, GET_POLLS } from "../actions/polls";

export function polls(state = [], action){
    switch(action.type){
        case GET_POLLS:
            console.log('action.polls:', action.polls);
            return action.polls;
        default: 
            return state;
    }
}
