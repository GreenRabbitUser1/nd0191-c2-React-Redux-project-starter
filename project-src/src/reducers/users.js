import { ADD_USER, GET_USERS } from "../actions/users";

export function users(state = [], action){
    switch(action.type){
        case GET_USERS:
            return action.users;
        default: 
            return state;
    }

}
