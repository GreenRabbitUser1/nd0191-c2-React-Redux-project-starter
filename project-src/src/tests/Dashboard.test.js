import React from "react";
import { render, cleanup } from "@testing-library/react";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import store from "../store";
import { _getQuestions, _getUsers } from "../../_DATA";
import { getUsers, saveUser } from "../actions/users";
import { getPolls } from "../actions/polls";
import Dashboard from "../components/Dashboard";

describe("Dashboard", () => {
    beforeAll(() => {
        Object.defineProperty(window, "matchMedia", {
            writable: true,
            value: jest
                .fn()
                .mockImplementation((query) => ({
                    matches: false,
                    media: query,
                    onchange: null,
                    addEventListener: jest.fn(),
                    removeEventListener: jest.fn(),
                    dispatchEvent: jest.fn(),
                    addListener: jest.fn(),
                    removeListener: jest.fn(),
                }))
        });
        delete window.matchMedia;
        window.matchMedia = (query) => ({
            matches: false,
            media: query,
            onchange: null,
            addListener: jest.fn(),
            removeListener: jest.fn(),
            dispatchEvent: jest.fn(),
            addEventListener: jest.fn(),
            removeEventListener: jest.fn(),
        });
    });
    afterEach(cleanup);

    it("should return a snapshot", async() => {
        const [users, questions] = await Promise.all([_getUsers(), _getQuestions()]);
        store.dispatch(getUsers(users));
        store.dispatch(getPolls(questions));
        store.dispatch(saveUser(store.getState().users["tylermcginnis"]));

        const component = render(
            <Provider store={store}>
                <BrowserRouter>
                    <Dashboard/>
                </BrowserRouter>
            </Provider>
        );

        expect(component).toBeDefined();
        expect(component).toMatchSnapshot();
    });
});