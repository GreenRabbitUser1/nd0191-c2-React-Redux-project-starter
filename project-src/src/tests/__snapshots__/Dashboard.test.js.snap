// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard Component should return a snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="view-wrap"
      >
        <div
          class="view-filter-top"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 512 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M256 105c-101.8 0-188.4 62.4-224 151 35.6 88.6 122.2 151 224 151s188.4-62.4 224-151c-35.6-88.6-122.2-151-224-151zm0 251.7c-56 0-101.8-45.3-101.8-100.7S200 155.3 256 155.3 357.8 200.6 357.8 256 312 356.7 256 356.7zm0-161.1c-33.6 0-61.1 27.2-61.1 60.4s27.5 60.4 61.1 60.4 61.1-27.2 61.1-60.4-27.5-60.4-61.1-60.4z"
            />
          </svg>
          <span>
            Filter View
          </span>
          <svg
            class="
                        view-filter-icon 
                        
                    "
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 448 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
            />
          </svg>
        </div>
      </div>
      <div
        class="polls"
      >
        <div
          class="poll-category"
        >
          <div
            class="poll-category-name"
          >
            Unanswered Polls
          </div>
          <div
            class="poll-wrap"
          />
        </div>
        <div
          class="poll-category"
        >
          <div
            class="poll-category-name"
          >
            Answered Polls
          </div>
          <div
            class="poll-wrap"
          />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="view-wrap"
    >
      <div
        class="view-filter-top"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M256 105c-101.8 0-188.4 62.4-224 151 35.6 88.6 122.2 151 224 151s188.4-62.4 224-151c-35.6-88.6-122.2-151-224-151zm0 251.7c-56 0-101.8-45.3-101.8-100.7S200 155.3 256 155.3 357.8 200.6 357.8 256 312 356.7 256 356.7zm0-161.1c-33.6 0-61.1 27.2-61.1 60.4s27.5 60.4 61.1 60.4 61.1-27.2 61.1-60.4-27.5-60.4-61.1-60.4z"
          />
        </svg>
        <span>
          Filter View
        </span>
        <svg
          class="
                        view-filter-icon 
                        
                    "
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
          />
        </svg>
      </div>
    </div>
    <div
      class="polls"
    >
      <div
        class="poll-category"
      >
        <div
          class="poll-category-name"
        >
          Unanswered Polls
        </div>
        <div
          class="poll-wrap"
        />
      </div>
      <div
        class="poll-category"
      >
        <div
          class="poll-category-name"
        >
          Answered Polls
        </div>
        <div
          class="poll-wrap"
        />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Dashboard should return a snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="view-wrap"
      >
        <div
          class="view-filter-top"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 512 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M256 105c-101.8 0-188.4 62.4-224 151 35.6 88.6 122.2 151 224 151s188.4-62.4 224-151c-35.6-88.6-122.2-151-224-151zm0 251.7c-56 0-101.8-45.3-101.8-100.7S200 155.3 256 155.3 357.8 200.6 357.8 256 312 356.7 256 356.7zm0-161.1c-33.6 0-61.1 27.2-61.1 60.4s27.5 60.4 61.1 60.4 61.1-27.2 61.1-60.4-27.5-60.4-61.1-60.4z"
            />
          </svg>
          <span>
            Filter View
          </span>
          <svg
            class="
                        view-filter-icon 
                        
                    "
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 448 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
            />
          </svg>
        </div>
      </div>
      <div
        class="polls"
      >
        <div
          class="poll-category"
        >
          <div
            class="poll-category-name"
          >
            Unanswered Polls
          </div>
          <div
            class="poll-wrap"
          />
        </div>
        <div
          class="poll-category"
        >
          <div
            class="poll-category-name"
          >
            Answered Polls
          </div>
          <div
            class="poll-wrap"
          />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="view-wrap"
    >
      <div
        class="view-filter-top"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M256 105c-101.8 0-188.4 62.4-224 151 35.6 88.6 122.2 151 224 151s188.4-62.4 224-151c-35.6-88.6-122.2-151-224-151zm0 251.7c-56 0-101.8-45.3-101.8-100.7S200 155.3 256 155.3 357.8 200.6 357.8 256 312 356.7 256 356.7zm0-161.1c-33.6 0-61.1 27.2-61.1 60.4s27.5 60.4 61.1 60.4 61.1-27.2 61.1-60.4-27.5-60.4-61.1-60.4z"
          />
        </svg>
        <span>
          Filter View
        </span>
        <svg
          class="
                        view-filter-icon 
                        
                    "
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
          />
        </svg>
      </div>
    </div>
    <div
      class="polls"
    >
      <div
        class="poll-category"
      >
        <div
          class="poll-category-name"
        >
          Unanswered Polls
        </div>
        <div
          class="poll-wrap"
        />
      </div>
      <div
        class="poll-category"
      >
        <div
          class="poll-category-name"
        >
          Answered Polls
        </div>
        <div
          class="poll-wrap"
        />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
